lib/execute.js

```diff
@@ -16,68 +16,149 @@
-  const filename = `${packageInfo.name}-${random}${extension}`;
-  return path.join(os.tmpdir(), filename);
+exports.getTemporaryScriptFilePath = (script) => {
+  const scriptFileName = `${packageJSON.name}-${packageJSON.version}-${script.originalFilename}`;
+  return path.join(os.tmpdir(), scriptFileName);
 };
 
 /**
- * @summary Create a temporary script file with the given contents
+ * @summary Create an executable file with certain text contents
  * @function
  * @private
  *
- * @param {String} extension - temporary script file extension
- * @param {String} contents - temporary script contents
- * @param {Function} callback - callback(error, temporaryPath)
+ * @param {String} content - string contents
+ * @param {String} output - output file path
+ * @param {Function} callback - callback (error)
  *
  * @example
- * createTemporaryScriptFile('.sh', '#!/bin/bash\necho "Foo"', (error, temporaryPath) => {
- *   if (error) {
- *     throw error;
- *   }
- *   console.log(temporaryPath);
- * })
+ * createExecutableFile('#!/bin/bash\necho "foo"', 'myscript.sh', (error) => {
+ *   if (error) throw error;
+ * });
  */
-const createTemporaryScriptFile = (extension, contents, callback) => {
-  const temporaryPath = tmpFilename(extension);
-  fs.writeFile(temporaryPath, contents, {
-    mode: 0o755
-  }, (error) => {
-    debug('write %s:', temporaryPath, error || 'OK');
-    callback(error, temporaryPath);
+const createExecutableFile = (content, output, callback) => {
+
+  // Lets cleanup the output location just in case its something
+  // that might prevent us from creating the file later on, like
+  // a directory with the exact same name.
+  rimraf(output, (error) => {
+
+    if (error) {
+      return callback(error);
+    }
+
+    return fs.writeFile(output, content, {
+      mode: MODE_EXECUTABLE
+    }, callback);
+  });
+};
+
+/**
+ * @summary Create a script file, caching when possible
+ * @function
+ * @private
+ *
+ * @param {Object} script - script object
+ * @param {String} output - output file path
+ * @param {Function} callback - callback (error)
+ *
+ * @example
+ * const scripts = require('./scripts.json');
+ * createScriptFile(scripts.darwin, 'myscript.sh', (error) => {
+ *   if (error) throw error;
+ * });
+ */
+const createScriptFile = (script, output, callback) => {
+  utils.isFile(output, (error, isFile) => {
+    if (error) {
+      return callback(error);
+    }
+
+    if (!isFile) {
+      debug('Creating script for the first time', output);
+      return createExecutableFile(script.content, output, callback);
+    }
+
+    // Integrity check, so we don't expose a vulnerability by letting
+    // other processes make this module run arbitrary code.
+    checksum.calculateFromFile(script.checksumType, output, (checksumError, scriptChecksum) => {
+      if (checksumError) {
+        return callback(checksumError);
+      }
+
+      if (script.checksum === scriptChecksum) {
```

Comment:
They all have the same prefix, so that's irrelevant too. Either way, if something is on your machine, it's game over already – not sure if this vector makes sense worrying about in the first place.